### ISSUE TITLE
Dispersy does not allow custom authentication, resolution, distribution, or destination methods

### ISSUE THREAD

**Issue by NielsZeilemaker** [Tribler/dispersy#145]
I'm trying to extend a fullsyncdistribution to only sync between friends.
In order to allow for this, i'm implementing a new distribution policy which extends fullsyncdistribution.

However, even replacing the fullsyncdistribution with a new class results in the following error:

``` python:
2013-10-29 11:59:10+0100 [-] KeyError: <class 'Tribler.community.privatesocial.destination.FOAFDestination'>
...
2013-10-29 11:59:10+0100 [-]     self._decode_message_map[byte] = self.DecodeFunctions(meta, mapping[type(meta.authentication)], mapping[type(meta.resolution)], mapping[type(meta.distribution)], mapping[type(meta.destination)], decode_payload_func)
```

Any idea how we can improve upon this, and add some flexibility to Dispersy without me needing to create my "own" dispersy variant which does support this destination policy?

**Comment by boudewijn-tribler** [comment]
I agree that it is currently to difficult to change/modify the behavior of policies.  Ideally it should use an OO approach, and have each message have functions that can be overridden (encode, decode, handle, create, etc) when needed.  Unfortunately this will have a large impact on the code.

As for the problem (syncing specific messages only between friends), it will ultimately be handled when a bloom filter is received.  Unless we support this in the Dispersy core, the community will need to do this by overriding the method handling incoming blooms (we can do this by moving this handler method from dispersy.py to community.py, see #100).

**Comment by NielsZeilemaker** [comment]
By replacing the handler i'm able to override default behaviour when a bloom is received. However, this is not the point of this issue. It's not being able to implement a authentication, resolution, distribution, or destination method without building your own Dispersy.
